csi.proto

```diff
@@ -1634,3 +1682,200 @@ message NodeExpandVolumeResponse {
   // The capacity of the volume in bytes. This field is OPTIONAL.
   int64 capacity_bytes = 1;
 }
+message GroupControllerGetCapabilitiesRequest {
+  // Intentionally empty.
+}
+
+message GroupControllerGetCapabilitiesResponse {
+  // All the capabilities that the group controller service supports.
+  // This field is OPTIONAL.
+  repeated GroupControllerServiceCapability capabilities = 1;
+}
+
+// Specifies a capability of the group controller service.
+message GroupControllerServiceCapability {
+  message RPC {
+    enum Type {
+      UNKNOWN = 0;
+
+      // Indicates that the group controller plugin supports
+      // creating and deleting a volume group snapshot.
+      CREATE_DELETE_VOLUME_GROUP_SNAPSHOT = 1
+      [(alpha_enum_value) = true];
+
+      // Indicates that the group controller plugin supports getting
+      // details of a volume group snapshot.
+      GET_VOLUME_GROUP_SNAPSHOT = 2 [(alpha_enum_value) = true];
+
+      // Indicates that the group controller plugin supports getting
+      // details of a list of volume group snapshots.
+      LIST_VOLUME_GROUP_SNAPSHOTS = 3 [(alpha_enum_value) = true];
+    }
+
+    Type type = 1;
+  }
+
+  oneof type {
+    // RPC that the controller supports.
+    RPC rpc = 1;
+  }
+}
+message CreateVolumeGroupSnapshotRequest {
+  option (alpha_message) = true;
+
+  // The suggested name for the group snapshot. This field is REQUIRED
+  // for idempotency.
+  // Any Unicode string that conforms to the length limit is allowed
+  // except those containing the following banned characters:
+  // U+0000-U+0008, U+000B, U+000C, U+000E-U+001F, U+007F-U+009F.
+  // (These are control characters other than commonly used whitespace.)
+  string name = 1;
+
+  // volume ids of the source volumes to be snapshotted together.
+  // This field is REQUIRED.
+  repeated string source_volume_ids = 2;
+
+  // Volume secrets required by plugin to complete volume group snapshot
+  // creation request.
+  // This field is OPTIONAL.
+  repeated VolumeSecret volume_secrets = 3;
```

Comment:
Thanks for adding this -- it partially addresses the concern around per-volume secrets. I think we may need a per-snapshot secret in addition, so the original secrets field probably should be added back in addition to this field.

Having both fields would enable drivers that just have one secret for all the volumes to pass it in the top level secrets rather than needing to repeat it per-volume. The trick is going to be what guidance we give to COs so that both styles can work.

I think we should take this opportunity to be more prescriptive about secrets for this new feature. As mentioned in #371 the proliferation of secrets on the CO side is out of control, and because this is a new feature, we have an opportunity to avoid just repeating the pattern of the past.

I would like to spell out how many secrets CSI expects the CO to store per-group-snapshot (and I also like to do this for volume and volumesnapshots, but not as part of this change) and how they should be mapped to RPC calls. For example, we should probably spell out that any secret (other than the list of volume secrets) supplied to this call should be the exact same secret supplied to Get and Delete for the same group snapshot, so the CO doesn't ever need to store more than 1 secret per group snapshot.

Reply:
I'm still at a loss for what to do about secrets in List calls. Have you considered leaving the List call out of this proposal entirely? List could always be added in the future if we find a specific use case.

Reply:
I added List only for completeness.  I can drop it for now.

Reply:
Added comments that secrets should be the same for the same group snapshot unless if the secrets are rotated.